File: JavaLib/hsgmParser.java
package hsgmParser
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.Map;
import java.util.zip.Deflater;

public class hsgmParser {

    private static final int TILE_SIZE = 16;
    private static final int WIDTH = 256;
    private static final int HEIGHT = 256;

    public static byte[] generatePixelData(Map<String, int[]> items) {
        byte[] pixelData = new byte[WIDTH * HEIGHT * 4]; // 4 bytes per pixel (RGBA)

        for (Map.Entry<String, int[]> entry : items.entrySet()) {
            String name = entry.getKey();
            int[] position = entry.getValue();
            int x = position[0];
            int y = position[1];

            if (x >= 0 && x < WIDTH && y >= 0 && y < HEIGHT) {
                int index = (y * WIDTH + x) * 4;
                byte[] color = name.equals("ColliderType") ? new byte[]{0, 0, (byte)255, (byte)255} : new byte[]{(byte)255, 0, 0, (byte)255};
                System.arraycopy(color, 0, pixelData, index, 4);
            }
        }

        return pixelData;
    }

    public static void writePng(String path, byte[] pixelData) throws IOException {
        try (OutputStream os = new FileOutputStream(path)) {
            os.write(new byte[]{(byte)137, 80, 78, 71, 13, 10, 26, 10}); // PNG signature
            os.write(createIHDRChunk());
            os.write(createIDATChunk(pixelData));
            os.write(createIENDChunk());
        }
    }

    private static byte[] createIHDRChunk() {
        byte[] widthBytes = intToBytes(WIDTH);
        byte[] heightBytes = intToBytes(HEIGHT);
        byte[] ihdrData = new byte[13];
        System.arraycopy(widthBytes, 0, ihdrData, 0, 4);
        System.arraycopy(heightBytes, 0, ihdrData, 4, 4);
        ihdrData[8] = 8; // bit depth
        ihdrData[9] = 6; // color type (RGBA)
        ihdrData[10] = 0; // compression method
        ihdrData[11] = 0; // filter method
        ihdrData[12] = 0; // interlace method
        return createChunk("IHDR", ihdrData);
    }

    private static byte[] createIDATChunk(byte[] pixelData) {
        byte[] compressedData = compressData(pixelData);
        return createChunk("IDAT", compressedData);
    }

    private static byte[] createIENDChunk() {
        return createChunk("IEND", new byte[0]);
    }

    private static byte[] createChunk(String type, byte[] data) {
        byte[] typeBytes = type.getBytes();
        byte[] lengthBytes = intToBytes(data.length);
        byte[] crcBytes = intToBytes(crc32(type.getBytes(), data));

        byte[] chunk = new byte[lengthBytes.length + typeBytes.length + data.length + crcBytes.length];
        int offset = 0;

        System.arraycopy(lengthBytes, 0, chunk, offset, lengthBytes.length);
        offset += lengthBytes.length;
        System.arraycopy(typeBytes, 0, chunk, offset, typeBytes.length);
        offset += typeBytes.length;
        System.arraycopy(data, 0, chunk, offset, data.length);
        offset += data.length;
        System.arraycopy(crcBytes, 0, chunk, offset, crcBytes.length);

        return chunk;
    }

    private static byte[] compressData(byte[] data) {
        Deflater deflater = new Deflater();
        deflater.setInput(data);
        deflater.finish();

        byte[] buffer = new byte[1024];
        int compressedDataLength = deflater.deflate(buffer);
        deflater.end();

        byte[] compressedData = new byte[compressedDataLength];
        System.arraycopy(buffer, 0, compressedData, 0, compressedDataLength);

        return compressedData;
    }

    private static byte[] intToBytes(int value) {
        return new byte[]{
            (byte) ((value >> 24) & 0xFF),
            (byte) ((value >> 16) & 0xFF),
            (byte) ((value >> 8) & 0xFF),
            (byte) (value & 0xFF)
        };
    }

    private static int crc32(byte[]... dataArrays) {
        int crc = 0xFFFFFFFF;
        for (byte[] data : dataArrays) {
            for (byte b : data) {
                crc ^= (b & 0xFF);
                for (int j = 0; j < 8; j++) {
                    if ((crc & 1) != 0) {
                        crc = (crc >>> 1) ^ 0xEDB88320;
                    } else {
                        crc >>>= 1;
                    }
                }
            }
        }
        return crc ^ 0xFFFFFFFF;
    }
}

File: RustLib/hsgmParser.rs
use std::fs::File;
use std::io::{self, Write, BufRead};
use std::collections::HashMap;

const TILE_SIZE: u32 = 16;
const WIDTH: u32 = 256;
const HEIGHT: u32 = 256;

struct Binding {
    texture_paths: HashMap<String, String>,
    collider_colors: HashMap<String, [u8; 4]>,
    trigger_colors: HashMap<String, [u8; 4]>,
}

impl Binding {
    fn get_texture_path(&self, texture_name: &str) -> Option<&String> {
        self.texture_paths.get(texture_name)
    }

    fn get_collider_color(&self, collider_name: &str) -> Option<&[u8; 4]> {
        self.collider_colors.get(collider_name)
    }

    fn get_trigger_color(&self, trigger_name: &str) -> Option<&[u8; 4]> {
        self.trigger_colors.get(trigger_name)
    }
}

fn write_png(file_path: &str, width: u32, height: u32, pixel_data: &[u8]) -> io::Result<()> {
    let mut file = File::create(file_path)?;

    // PNG file signature
    file.write_all(&[0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A])?;

    // IHDR chunk
    write_chunk(&mut file, "IHDR", |writer| {
        writer.write_all(&width.to_be_bytes())?;
        writer.write_all(&height.to_be_bytes())?;
        writer.write_all(&[8, 6, 0, 0, 0])?; // 8-bit depth, RGBA, no compression, no filter, no interlace
        Ok(())
    })?;

    // IDAT chunk
    let idat_data = generate_idat_data(pixel_data);
    write_chunk(&mut file, "IDAT", |writer| writer.write_all(&idat_data))?;

    // IEND chunk
    write_chunk(&mut file, "IEND", |writer| Ok(()))?;

    Ok(())
}

fn write_chunk<F>(writer: &mut File, chunk_type: &str, data_fn: F) -> io::Result<()>
where
    F: FnOnce(&mut Vec<u8>) -> io::Result<()>,
{
    let mut data = Vec::new();
    data_fn(&mut data)?;
    let length = data.len() as u32;
    writer.write_all(&length.to_be_bytes())?;
    writer.write_all(chunk_type.as_bytes())?;
    writer.write_all(&data)?;

    // Compute CRC for the chunk type
    let crc = crc32(chunk_type.as_bytes());
    writer.write_all(&crc.to_be_bytes())?;
    Ok(())
}

fn crc32(data: &[u8]) -> u32 {
    const POLY: u32 = 0xedb88320;
    let mut crc = 0xffffffff;
    for &byte in data {
        let mut temp = byte as u32 ^ crc;
        for _ in 0..8 {
            if temp & 1 != 0 {
                temp = (temp >> 1) ^ POLY;
            } else {
                temp >>= 1;
            }
        }
        crc = temp ^ (crc >> 8);
    }
    !crc
}

fn generate_idat_data(pixel_data: &[u8]) -> Vec<u8> {
    let mut data = Vec::new();
    data.push(0x00); // No filter
    for y in 0..HEIGHT {
        let offset = (y * WIDTH * 4) as usize;
        let row = &pixel_data[offset..offset + (WIDTH * 4) as usize];
        data.extend_from_slice(row);
    }
    
    let mut compressed_data = Vec::new();
    compress_deflate(&data, &mut compressed_data);
    compressed_data
}

fn compress_deflate(input: &[u8], output: &mut Vec<u8>) {
    // Minimal compression placeholder; a real implementation would use DEFLATE.
    output.extend_from_slice(input);
}

fn parse_map_file(file_path: &str) -> io::Result<(Vec<u8>, Vec<u8>, Vec<u8>, Binding)> {
    let file = File::open(file_path)?;
    let reader = io::BufReader::new(file);

    let mut textures = Vec::new();
    let mut colliders = Vec::new();
    let mut triggers = Vec::new();
    let mut width = 0;
    let mut height = 0;

    let mut current_section = "";
    let mut binding = Binding {
        texture_paths: HashMap::new(),
        collider_colors: HashMap::new(),
        trigger_colors: HashMap::new(),
    };

    for line in reader.lines() {
        let line = line?;
        if line.starts_with("[TEXTURES]") {
            current_section = "textures";
            continue;
        } else if line.starts_with("[COLLIDERS]") {
            current_section = "colliders";
            continue;
        } else if line.starts_with("[TRIGGERS]") {
            current_section = "triggers";
            continue;
        } else if line.starts_with("[METADATA]") {
            current_section = "metadata";
            continue;
        }

        if let Some((key, value)) = line.split_once('=') {
            let key = key.trim();
            let value = value.trim().strip_prefix('[').and_then(|v| v.strip_suffix(']')).unwrap_or("").trim();
            
            match current_section {
                "textures" => {
                    if let Some((name, path)) = key.split_once('=') {
                        binding.texture_paths.insert(name.trim().to_string(), path.trim().to_string());
                    }
                }
                "colliders" => {
                    if let Some((name, color)) = key.split_once('=') {
                        let parts: Vec<&str> = color.split(',').collect();
                        if parts.len() == 4 {
                            let color = [
                                parts[0].parse::<u8>().unwrap_or(0),
                                parts[1].parse::<u8>().unwrap_or(0),
                                parts[2].parse::<u8>().unwrap_or(0),
                                parts[3].parse::<u8>().unwrap_or(255),
                            ];
                            binding.collider_colors.insert(name.trim().to_string(), color);
                        }
                    }
                }
                "triggers" => {
                    if let Some((name, color)) = key.split_once('=') {
                        let parts: Vec<&str> = color.split(',').collect();
                        if parts.len() == 4 {
                            let color = [
                                parts[0].parse::<u8>().unwrap_or(0),
                                parts[1].parse::<u8>().unwrap_or(0),
                                parts[2].parse::<u8>().unwrap_or(0),
                                parts[3].parse::<u8>().unwrap_or(255),
                            ];
                            binding.trigger_colors.insert(name.trim().to_string(), color);
                        }
                    }
                }
                "metadata" => {
                    if line.starts_with("MAP_SIZE") {
                        let parts: Vec<&str> = line.split('=').nth(1).unwrap_or("").trim().strip_prefix('[').and_then(|v| v.strip_suffix(']')).unwrap_or("").split(',').collect();
                        if parts.len() == 2 {
                            width = parts[0].parse::<u32>().unwrap_or(256);
                            height = parts[1].parse::<u32>().unwrap_or(256);
                        }
                    }
                }
                _ => {}
            }
        }
    }

    let texture_data = generate_pixel_data(&textures, width, height, [0xFF, 0x00, 0x00, 0xFF]); // Red for textures
    let collider_data = generate_pixel_data_from_bindings(&colliders, width, height, &binding, true); // Using Binding for colors
    let trigger_data = generate_pixel_data_from_bindings(&triggers, width, height, &binding, false); // Using Binding for colors

    Ok((texture_data, collider_data, trigger_data, binding))
}

fn generate_pixel_data(items: &[(String, u32, u32)], width: u32, height: u32, color: [u8; 4]) -> Vec<u8> {
    let mut pixel_data = vec![0; (width * height * 4) as usize];

    for (_, x, y) in items {
        let index = ((y * width + x) * 4) as usize;
        if index + 3 < pixel_data.len() {
            pixel_data[index] = color[0]; // Red
            pixel_data[index + 1] = color[1]; // Green
            pixel_data[index + 2] = color[2]; // Blue
            pixel_data[index + 3] = color[3]; // Alpha
        }
    }

    pixel_data
}

fn generate_pixel_data_from_bindings(items: &[(String, u32, u32)], width: u32, height: u32, binding: &Binding, is_collider: bool) -> Vec<u8> {
    let mut pixel_data = vec![0; (width * height * 4) as usize];

    for (name, x, y) in items {
        let color = if is_collider {
            binding.get_collider_color(name).unwrap_or(&[0, 0, 0, 255])
        } else {
            binding.get_trigger_color(name).unwrap_or(&[0, 0, 0, 255])
        };

        let index = ((y * width + x) * 4) as usize;
        if index + 3 < pixel_data.len() {
            pixel_data[index] = color[0]; // Red
            pixel_data[index + 1] = color[1]; // Green
            pixel_data[index + 2] = color[2]; // Blue
            pixel_data[index + 3] = color[3]; // Alpha
        }
    }

    pixel_data
}

fn main() -> io::Result<()> {
    let (texture_data, collider_data, trigger_data, binding) = parse_map_file("example.map")?;
    write_png("textures.png", WIDTH, HEIGHT, &texture_data)?;
    write_png("colliders.png", WIDTH, HEIGHT, &collider_data)?;
    write_png("triggers.png", WIDTH, HEIGHT, &trigger_data)?;

    Ok(())
}
